lsm/compaction.go

```diff
@@ -0,0 +1,852 @@
+package lsm
+
+import (
+	"bytes"
+	"container/heap"
+	"encoding/binary"
+	"fmt"
+	"github.com/google/uuid"
+	"github.com/spirit-labs/tektite/asl/arista"
+	"github.com/spirit-labs/tektite/asl/encoding"
+	"github.com/spirit-labs/tektite/common"
+	log "github.com/spirit-labs/tektite/logger"
+	"github.com/spirit-labs/tektite/sst"
+	"math"
+	"strings"
+	"time"
+)
+
+type compactionState struct {
+	tableDeleteTimer   *time.Timer
+	tablesToDelete     []deleteTableEntry
+	jobQueue           []jobHolder
+	inProgress         map[string]inProgressCompaction
+	pendingCompactions map[int]int
+	lockedRanges       map[int][]lockedRange
+	pollers            *pollerQueue
+	stats              CompactionStats
+}
+
+func newCompactionState() compactionState {
+	return compactionState{
+		inProgress:         make(map[string]inProgressCompaction),
+		pendingCompactions: map[int]int{},
+		lockedRanges:       map[int][]lockedRange{},
+		pollers:            &pollerQueue{},
+	}
+}
+
+type jobHolder struct {
+	job            CompactionJob
+	completionFunc func(error)
+}
+
+type lockedRange struct {
+	level int
+	start []byte
+	end   []byte
+}
+
+func (lr *lockedRange) overlaps(rng *lockedRange) bool {
+	dontOverlapRight := bytes.Compare(rng.start, lr.end) > 0
+	dontOverlapLeft := bytes.Compare(rng.end, lr.start) < 0
+	dontOverlap := dontOverlapLeft || dontOverlapRight
+	return !dontOverlap
+}
+
+func (lm *Manager) maybeScheduleCompaction() error {
+	// Get a level to compact (if any)
+	level, numTables := lm.chooseLevelToCompact()
```

Comment:
high level architecture q - Could two agents independently be compacting the same level at once? Would that waste cycles?

afaict yes. would this be avoidable with an s3 conditional write?

Reply:
Compaction only occurs on the agent that has the LSM controller for the shard.